r/R/dplyr-funcs-type.R

```diff
@@ -0,0 +1,250 @@
+        character = call_binding("is.character", object),
+        numeric = call_binding("is.numeric", object),
+        integer = call_binding("is.integer", object),
+        integer64 = call_binding("is.integer64", object),
+        logical = call_binding("is.logical", object),
+        factor = call_binding("is.factor", object),
+        list = call_binding("is.list", object),
+        # for Arrow data types, compare class2 with object$type()$ToString(),
+        # but first strip off any parameters to only compare the top-level data
+        # type,  and canonicalize class2
+        sub("^([^([<]+).*$", "\\1", object$type()$ToString()) ==
+          canonical_type_str(class2)
+      )
+    } else if (inherits(class2, "DataType")) {
+      object$type() == as_type(class2)
+    } else {
+      stop("Second argument to is() is not a string or DataType", call. = FALSE)
+    }
+  })
+
+  # Create a data frame/tibble/struct column
+  register_binding("tibble", function(..., .rows = NULL, .name_repair = NULL) {
+    if (!is.null(.rows)) arrow_not_supported(".rows")
+    if (!is.null(.name_repair)) arrow_not_supported(".name_repair")
+
+    # use dots_list() because this is what tibble() uses to allow the
+    # useful shorthand of tibble(col1, col2) -> tibble(col1 = col1, col2 = col2)
+    # we have a stronger enforcement of unique names for arguments because
+    # it is difficult to replicate the .name_repair semantics and expanding of
+    # unnamed data frame arguments in the same way that the tibble() constructor
+    # does.
+    args <- rlang::dots_list(..., .named = TRUE, .homonyms = "error")
+
+    build_expr(
+      "make_struct",
+      args = unname(args),
+      options = list(field_names = names(args))
+    )
+  })
+
+  register_binding("data.frame", function(..., row.names = NULL,
+                                              check.rows = NULL, check.names = TRUE, fix.empty.names = TRUE,
+                                              stringsAsFactors = FALSE) {
+    # we need a specific value of stringsAsFactors because the default was
+    # TRUE in R <= 3.6
+    if (!identical(stringsAsFactors, FALSE)) {
+      arrow_not_supported("stringsAsFactors = TRUE")
+    }
+
+    # ignore row.names and check.rows with a warning
+    if (!is.null(row.names)) arrow_not_supported("row.names")
+    if (!is.null(check.rows)) arrow_not_supported("check.rows")
+
+    args <- rlang::dots_list(..., .named = fix.empty.names)
+    if (is.null(names(args))) {
+      names(args) <- rep("", length(args))
+    }
+
+    if (identical(check.names, TRUE)) {
+      if (identical(fix.empty.names, TRUE)) {
+        names(args) <- make.names(names(args), unique = TRUE)
+      } else {
+        name_emtpy <- names(args) == ""
+        names(args)[!name_emtpy] <- make.names(names(args)[!name_emtpy], unique = TRUE)
+      }
+    }
+
+    build_expr(
+      "make_struct",
+      args = unname(args),
+      options = list(field_names = names(args))
+    )
+  })
+}
```

Comment:
```suggestion
```

This is extraneous, no? And it looks like the lines above this are also indented?

Reply:
Ah no, my bad, this actually needed `register_bindings_type_cast <- function() {` above it